### ISSUE TITLE
utils_format_graphite.c translates colons to dots in collectd 5.5

### ISSUE THREAD

**Issue by zerthimon** [collectd/collectd#1102]
In 5.5 amqp plugin sends graphite metric names differently than 5.4:
If the plugin instance name contains a colon (:), it's translated into a dot (.) This presents a problem as dots in metric name have special meaning in grapihte (they are used as separators).
As a result, I have a broken metric tree in graphite after upgrading to 5.5.

Works well in collectd:
```
# collectdctl -s /var/run/collectd.sock putval host/plugin-pl:inst/gauge N:1
# collectdctl -s /var/run/collectd.sock listval | grep host
receive: <-- 1435674171.443 host/plugin-pl:inst/gauge
host/plugin-pl:inst/gauge
#
```

Broken in graphite:
```
host.plugin-pl.inst.gauge
```

**Comment by zerthimon** [comment]
ping ?

**Comment by octo** [comment]
This is actually a change in the *write_graphite plugin*, introduced in fd4bdc17fc1560ed0775e54d01f564345ecea92e.

@pyr can you double-check if the colon is really a forbidden character by graphite?

**Comment by zerthimon** [comment]
This is probably due to format 'graphite' (/src/utils_format_graphite.c) not the AMQP plugin directly.

Especially this line https://github.com/collectd/collectd/blob/master/src/utils_format_graphite.c#L31
Why is semicolon a forbidden character in graphite ? Can we have it removed from this list ?

**Comment by zerthimon** [comment]
Also, as the dot has a special meaning in graphite (is used as field separator) translating 'forbidden' characters to a dot is not a good solution (imho) they should be translated into some 'neutral' character instead (underscore ?)

**Comment by octo** [comment]
@zerthimon the default escape character is `_`, a dot is really not the best choice.

**Comment by zerthimon** [comment]
Regarding so called 'forbidden' characters: Graphite doesn't really care about metric names as long as the character isn't a dot or some character that can mess up the FS.
Graphite (carbon) storage engine whisper (ceres ?) stores metrics as files which path is a metric name where dots are translated to slashes. So as long as a char cannot harm the directory order - it's OK. IMHO.

**Comment by zerthimon** [comment]
I'm gonna patch utils_format_graphite.c locally to drop colon from "forbidden" character list and build a package for my systems in a hope that this issue will be dealt with some day....

**Comment by mfournier** [comment]
If I remember correctly, carbon (the storage system) doesn't care about metric names, as long as the characters are considered valid by the underlying filesystem. But graphite (the web interface) doesn't handle some characters properly.

So what collectd does by default is replacing them with an underscore (`_`). write_graphite comes with the `EscapeCharacter` option, which lets the user change this default to any other character, including the dot (`.`), which would lead to the behaviour described here.

So as far as I can tell, this is more of a misconfiguration/misunderstanding than anything else.

**Comment by zerthimon** [comment]
@mfournier 
In my use case write_graphite plugin is not used. Also, EscapeCharacter is not defined (I have plugins that were written with with graphite naming scheme in mind and contain dots in metric name). The AMQP plugin is used to send metrics to carbon.
I'm sending metrics that contain ':' in it's name which is getting replaced with '.'
I have patched my utils_format_graphite.c and removed ':' from forbidden characters. There are NO PROBLEMS with this character in the web frontend (graphite). So why shoud ':' be banned from metric names ?

**Comment by zerthimon** [comment]
IMHO, having list of "Forbidden Characters" hardcoded is not the best approach. Instead, this should be made a Config parameter.

**Comment by mfournier** [comment]
Right, I read too quickly and missed out you were sending data to graphite using the amqp plugin, sorry for the confusion !

I'm pretty sure there was an issue with "`:`" characters at some point (maybe fixed in current graphite version ?), but I can't remember what :-( I doubt @pyr would have added it for no reason neither.

Whatever, I agree there is a need to make this configurable. Maybe also something more generic than just targeted to graphite (cf #876).

Thanks for bringing this forward @zerthimon !

**Comment by zerthimon** [comment]
Just ping....
Can we have this implemented in 5.7 ?
